windows/src/desktop/kmshell/main/Keyman.System.RemoteUpdateCheck.pas

```diff
@@ -0,0 +1,386 @@
+        Inc(FDownload.TotalDownloads);
+        Inc(FDownload.TotalSize, Params.Packages[i].DownloadSize);
+        Params.Packages[i].SavePath := SavePath + Params.Packages[i].FileName;
+      end;
+
+    // Add the Keyman installer
+    Inc(FDownload.TotalDownloads);
+    Inc(FDownload.TotalSize, Params.InstallSize);
+
+    // Keyboard Packages
+    FDownload.StartPosition := 0;
+    for i := 0 to High(Params.Packages) do
+      begin
+        if not DownloadFile(Params.Packages[i].DownloadURL, Params.Packages[i].SavePath) then // I2742
+        begin
+          Params.Packages[i].Install := False; // Download failed but install other files
+        end
+        else
+          Inc(downloadCount);
+        FDownload.StartPosition := FDownload.StartPosition + Params.Packages[i].DownloadSize;
+      end;
+
+    // Keyamn Installer
+    if not DownloadFile(Params.InstallURL, SavePath + Params.FileName) then  // I2742
+    begin
+      // TODO record fail? and log  // Download failed but user wants to install other files
+    end
+    else
+    begin
+      Inc(downloadCount)
+    end;
+
+    // There needs to be at least one file successfully downloaded to return
+    // TRUE that files where downloaded
+    if downloadCount > 0  then
+      Result := True;
+  except
+    on E:EHTTPUploader do
+    begin
+      if (E.ErrorCode = 12007) or (E.ErrorCode = 12029)
+        then LogMessage(S_OnlineUpdate_UnableToContact)
+        else LogMessage(WideFormat(S_OnlineUpdate_UnableToContact_Error, [E.Message]));
+      Result := False;
+    end;
+  end;
+end;
+
+function TRemoteUpdateCheck.DownloadUpdates(Params: TUpdateCheckResponse): Boolean;
+var
+  DownloadBackGroundSavePath : String;
+  DownloadResult : Boolean;
+begin
+  DownloadBackGroundSavePath := IncludeTrailingPathDelimiter(TKeymanPaths.KeymanUpdateCachePath);
+
+  DoDownloadUpdates(DownloadBackGroundSavePath, Params, DownloadResult);
+  KL.Log('TRemoteUpdateCheck.DownloadUpdatesBackground: DownloadResult = '+IntToStr(Ord(DownloadResult)));
+  Result := DownloadResult;
+
+end;
+
+function TRemoteUpdateCheck.DoRun: TRemoteUpdateCheckResult;
+var
+  flags: DWord;
+  i: Integer;
+  ucr: TUpdateCheckResponse;
+  pkg: IKeymanPackage;
+  downloadResult: boolean;
+begin
+  {FProxyHost := '';
+  FProxyPort := 0;}
+
+  { Check if user is currently online }
+  if not InternetGetConnectedState(@flags, 0) then
+  begin
+    Result := wucOffline;
+    Exit;
+  end;
+
+  { Verify that it has been at least 7 days since last update check }
+  try
+    with TRegistryErrorControlled.Create do  // I2890
+    try
+      if OpenKeyReadOnly(SRegKey_KeymanDesktop_CU) then
+      begin
+        if ValueExists(SRegValue_CheckForUpdates) and not ReadBool(SRegValue_CheckForUpdates) and not FForce then
+        begin
+          Result := wucNoUpdates;
+          Exit;
+        end;
+        if ValueExists(SRegValue_LastUpdateCheckTime) and (Now - ReadDateTime(SRegValue_LastUpdateCheckTime) < 7) and not FForce then
```

Comment:
Can we make this `7` a constant declaration at the top of the file?

Reply:
Already done in the next PR :)